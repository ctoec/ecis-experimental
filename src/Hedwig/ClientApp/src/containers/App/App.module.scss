// Arbitrary amount of space to temporarily avoid issue of things going off the page
// TODO: revisit this when we add a footer
.container {
	padding-bottom: 10rem;
}
