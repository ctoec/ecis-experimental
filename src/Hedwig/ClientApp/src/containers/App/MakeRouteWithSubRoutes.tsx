import React from 'react';
import { Route } from 'react-router-dom';
import { RouteConfig } from '../../routes';
import NotSignedInRedirect from '../NotSignedInRedirect/NotSignedInRedirect';

// Derived from: https://www.freecodecamp.org/news/hitchhikers-guide-to-react-router-v4-c98c39892399/

const MakeRouteWithSubRoutes = (route: RouteConfig) => {
	return (
		<Route
			path={route.path}
			exact={route.exact}
			render={(props) => {
				const component = <route.component {...props} {...route.props} routes={route.routes} />;
				if (route.unauthorized) {
					return component;
				} else {
					return <NotSignedInRedirect>{component}</NotSignedInRedirect>;
				}
			}}
		/>
	);
};

export default MakeRouteWithSubRoutes;
