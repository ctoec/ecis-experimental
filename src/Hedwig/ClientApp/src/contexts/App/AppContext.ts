import { createContext, useState } from 'react';

export type AppContextType = { 
	cacheInvalidator: number,
	invalidateCache: () => void
};

const AppContext = createContext<AppContextType>({
	cacheInvalidator: 0,
	invalidateCache: () => {}
});

const { Provider, Consumer } = AppContext;

export const useCacheInvalidator: () => AppContextType = () => {
	const [cacheInvalidator, setCacheInvalidator] = useState(0);
  const invalidateCache = () => setCacheInvalidator((prev) => prev + 1);
	return {cacheInvalidator, invalidateCache};
}

export { Provider as AppProvider };
export { Consumer as AppConsumer };
export default AppContext;