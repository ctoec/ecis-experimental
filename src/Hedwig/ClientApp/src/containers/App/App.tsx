import React, { useContext } from 'react';
import 'react-dates/initialize';
import { Switch } from 'react-router-dom';
import MakeRouteWithSubRoutes from './MakeRouteWithSubRoutes';
import routes from '../../routes';
import useApi from '../../hooks/useApi';
import { ApiOrganizationsOrgIdReportsGetRequest, CdcReport as Report } from '../../generated';
import { getIdForUser } from '../../utils/models';
import UserContext from '../../contexts/User/UserContext';
import { useCacheInvalidator, AppProvider } from '../../contexts/App/AppContext';
import { AlertProvider } from '../../contexts/Alert/AlertContext';
import { DeepNonUndefineable } from '../../utils/types';
import { ErrorBoundary, Header, NavItemProps } from '../../components';

import cx from 'classnames';
import styles from './App.module.scss';

/**
 * Main React component container for Hedwig application
 */
const App: React.FC = () => {
	const { user } = useContext(UserContext);
	const cacheContext = useCacheInvalidator();
	const { cacheInvalidator } = cacheContext;

	const params: ApiOrganizationsOrgIdReportsGetRequest = {
		orgId: getIdForUser(user, 'org'),
	};

	const [loading, error, reports] = useApi(
		api => api.apiOrganizationsOrgIdReportsGet(params),
		[user, cacheInvalidator],
		{
			skip: !user,
		}
	);

	const [, , organization] = useApi(
		api =>
			api.apiOrganizationsIdGet({
				id: getIdForUser(user, 'org'),
			}),
		[user],
		{ skip: !user }
	);

	const pendingReportsCount =
		!loading &&
		!error &&
		reports &&
		reports.filter<DeepNonUndefineable<Report>>(
			(r => !r.submittedAt) as (_: DeepNonUndefineable<Report>) => _ is DeepNonUndefineable<Report>
		).length;

	let navItems: NavItemProps[] = [
		{ type: 'secondary', title: 'Privacy policy', path: '/privacy-policy' },
		// { type: 'secondary', title: 'Help', path: '/help' },
	];

	if (user) {
		navItems = navItems.concat([
			{ type: 'primary', title: 'Roster', path: '/roster' },
			{
				type: 'primary',
				title:
					pendingReportsCount && pendingReportsCount > 0
						? `Reports (${pendingReportsCount})`
						: 'Reports',
				path: '/reports',
			},
		]);
	}

	return (
		<div className={cx(styles.container)}>
			<AppProvider value={cacheContext}>
				<a className="usa-skipnav" href="#main-content">
					Skip to main content
				</a>
				<Header
					primaryTitle="ECE Reporter"
					secondaryTitle={(organization && organization.name) || undefined}
					navItems={navItems}
					loginPath="/login"
					logoutPath="/logout"
					userFirstName={(user && user.firstName) || undefined}
				/>
				<main id="main-content">
					<ErrorBoundary>
						<AlertProvider>
							<Switch>
								{routes.map((route, index) => (
									<MakeRouteWithSubRoutes key={index} {...route} />
								))}
							</Switch>
						</AlertProvider>
					</ErrorBoundary>
				</main>
			</AppProvider>
		</div>
	);
};

export default App;
