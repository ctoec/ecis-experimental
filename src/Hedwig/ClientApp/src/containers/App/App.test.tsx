import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import App from './App';

it('renders without crashing', () => {
	const div = document.createElement('div');
	ReactDOM.render(
		<BrowserRouter>
			<App />
		</BrowserRouter>,
		div
	);
	ReactDOM.unmountComponentAtNode(div);
});

it('matches snapshot', () => {
	const { asFragment } = render(
		<MemoryRouter>
			<App />
		</MemoryRouter>
	);

	expect(asFragment()).toMatchSnapshot();
});
