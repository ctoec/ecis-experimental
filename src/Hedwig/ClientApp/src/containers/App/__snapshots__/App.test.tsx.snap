// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <a
      class="usa-skipnav"
      href="#main-content"
    >
      Skip to main content
    </a>
    <div
      class=""
    >
      <div
        class="usa-overlay"
        data-testid="overlay"
      />
      <header
        class="usa-header usa-header--extended usa-header--oec-default"
        role="banner"
      >
        <div
          class="usa-navbar"
        >
          <div
            class="usa-logo"
            id="extended-logo"
          >
            <a
              aria-label="ECE Reporter home"
              href="/"
            >
              <div
                class="usa-logo__text display-flex flex-align-center"
              >
                <div>
                  <div
                    class="primary-title primary-title--only"
                  >
                    ECE Reporter
                  </div>
                </div>
              </div>
            </a>
          </div>
          <button
            class="usa-menu-btn"
          >
            Menu
          </button>
        </div>
        <nav
          class="usa-nav"
          role="navigation"
        >
          <div
            class="usa-nav__inner"
          >
            <button
              class="usa-nav__close"
            >
              <img
                alt="close"
                src="close.svg"
              />
            </button>
            <ul
              class="usa-nav__primary usa-accordion"
            />
            <div
              class="usa-nav__secondary usa-nav__secondary--extended"
            >
              <ul
                class="usa-nav__secondary-links"
              >
                <li
                  class="usa-nav__secondary-item"
                >
                  <a
                    href="/privacy-policy"
                  >
                    Privacy policy
                  </a>
                </li>
              </ul>
              <div
                class="oec-logged-in-user"
              >
                <a
                  href="/login"
                >
                  Sign in
                </a>
              </div>
            </div>
          </div>
        </nav>
      </header>
    </div>
    <main
      class=""
      id="main-content"
    />
  </div>
</DocumentFragment>
`;
